File: Medium/BinaryTreeLevelOrderTraversal.cpp
//https://leetcode.com/problems/binary-tree-level-order-traversal/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(root==nullptr){
            return {};
        }
        std::vector<vector<int>> answer;
        answer.push_back({root->val});

        std::vector<TreeNode*> currentLevel;
        currentLevel.push_back(root);

        while(currentLevel.size()>0){
            updateLevel(currentLevel, answer);
        }
        return answer;
    }

    void updateLevel(vector<TreeNode*> &currentLevel, vector<vector<int>> &answer){
        std::vector<int> values;
        int startSize = currentLevel.size();
        for(int i = 0; i < startSize; i++){
            if(currentLevel[i]->left != nullptr){
                currentLevel.push_back(currentLevel[i]->left);
                values.push_back(currentLevel[i]->left->val);
            }
            if(currentLevel[i]->right != nullptr){
                currentLevel.push_back(currentLevel[i]->right);
                values.push_back(currentLevel[i]->right->val);
            }
        }

        currentLevel.erase(currentLevel.begin(), currentLevel.begin() + startSize);

        if(values.size()>0){
            answer.push_back(values);
        }
    }
};

File: Medium/BinaryTreeLevelOrderTraversal/BinaryTreeLevelOrderTraversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # VALID DFS SOLUTION
        # if not root:
        #     return []

        # result = []

        # def dfs(node:Optional[TreeNode], level:int, result:List[int]) -> None:
        #     if node is None:
        #         return

        #     if level>=len(result):
        #         result.append([])
        #     result[level].append(node.val)

        #     dfs(node.left, level+1, result)
        #     dfs(node.right, level+1, result)

        # dfs(root, 0, result)
        # return result



        #BFS SOLUTION
        if not root:
            return []

        result = []
        q = deque([root])

        while q:
            curr = []
            for i in range(len(q)):
                node = q.popleft()
                curr.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
                
            result.append(curr)

        return result
